src/case_setup/Poisson/poisson.f90

```diff
@@ -197,53 +205,69 @@ program poisson
   call initialise(mat_properties)
   call initialise(poisson_eq)
 
-  ! Create stiffness matrix
   call set_size(par_env, mesh, mat_properties)
   call set_nnz(5, mat_properties)
-  call create_matrix(mat_properties, M)
-
-  call discretise_poisson(mesh, M)
-
-  call begin_update(M) ! Start the parallel assembly for M
 
   ! Create right-hand-side and solution vectors
   call set_size(par_env, mesh, vec_properties)
-  call create_vector(vec_properties, b)
   call create_vector(vec_properties, u_exact)
-  call create_vector(vec_properties, u)
+  call create_vector(vec_properties, diag)
+  call create_vector(vec_properties, u%values)
 
-  call begin_update(u) ! Start the parallel assembly for u
+  ! Initialise with exact solution
+  call set_exact_sol(u_exact)
 
-  ! Evaluate right-hand-side vector
-  call eval_rhs(mesh, b)
+  call update(u%values)
+  call initialise_old_values(vec_properties, u)
 
-  call begin_update(b) ! Start the parallel assembly for b
-  call end_update(M) ! Complete the parallel assembly for M
-  call end_update(b) ! Complete the parallel assembly for b
+  call activate_timestepping()
+  call set_timestep(dt)
 
-  ! Modify matrix and right-hand-side vector to apply Dirichlet boundary conditions
-  call apply_dirichlet_bcs(mesh, M, b)
-  call begin_update(b) ! Start the parallel assembly for b
-  call finalise(M)
+  do t = 1, num_steps
 
-  call end_update(u) ! Complete the parallel assembly for u
-  call end_update(b) ! Complete the parallel assembly for b
+    call update_old_values(u)
 
-  ! Create linear solver & set options
-  call set_equation_system(par_env, b, u, M, poisson_eq)
-  call create_solver(poisson_eq, poisson_solver)
-  call set_solver_method(pressure_solver_method_name, poisson_solver)
-  call set_solver_precon(pressure_solver_precon_name, poisson_solver)
-  call solve(poisson_solver)
+    ! Create stiffness matrix
+    call create_matrix(mat_properties, M)
+    call create_vector(vec_properties, b)
```

Comment:
I think we can do this outside the time loop?

Reply:
I get the following error at the second timestep when doing so `mat_petsc.f90(181)[0] : ERROR: changing matrix mode without updating`
So I am going to leave it there for now